public/apps/configuration/panels/audit-logging/code-editor.tsx

```diff
@@ -0,0 +1,74 @@
+/*
+ *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+import { EuiCodeEditor, EuiText, EuiTextColor } from '@elastic/eui';
+import React, { useState } from 'react';
+import { get } from 'lodash';
+import { AuditLoggingSettings } from './types';
+import { SettingContent } from './constants';
+import { displayObject } from '../../utils/display-utils';
+import 'brace/theme/textmate';
+import 'brace/mode/json';
+
+export function CodeEditor(props: {
+  setting: SettingContent;
+  config: AuditLoggingSettings;
+  handleChange: (setting: SettingContent, val: boolean | string[]) => void;
+  handleInvalid: (key: string, error: boolean) => void;
+}) {
```

Comment:
I would suggest make the component more closure-like. Means not involving SettingContent or AuditLoggingSettings just pass a value, handleChange(val), handleInvalid(error), and do the encapsulation at upper layer.

In this case:
1. It does has the ability to access a field doesn't belong to it.
2. Easier to do unit testing.

Reply:
This CodeEditor already encapsulates the native EuiCodeEditor which takes the simple const onChange = (newValue: string) function which is agnostic about SettingContent or AuditLoggingSettings. We need a place to perform the value validation and decide to update the states of its caller. Not sure where you suggest to push this layer of logic further up to.

Reply:
I see. Fixed it. The updated JsonCodeEditor is much cleaner. Good suggestion!